fedimint-api/src/db/mod.rs

```diff
@@ -56,16 +57,83 @@ pub trait DatabaseValue: Sized + SerializableDatabaseValue {
 }
 
 pub type PrefixIter<'a> = Box<dyn Iterator<Item = Result<(Vec<u8>, Vec<u8>)>> + Send + 'a>;
+type Notifications = DashMap<Box<[u8]>, Arc<Notify>>;
 
 #[async_trait]
 pub trait IDatabase: Debug + Send + Sync {
-    async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction;
+    async fn begin_transaction<'a>(&'a self) -> Box<dyn IDatabaseTransaction<'a> + Send + 'a>;
 }
 
-dyn_newtype_define! {
-    /// A handle to a type-erased database implementation
-    #[derive(Clone)]
-    pub Database(Arc<IDatabase>)
+#[derive(Clone, Debug)]
+pub struct Database {
+    inner: Arc<dyn IDatabase>,
+    notifications: Arc<Notifications>,
+}
+
+impl<D: IDatabase + 'static> From<D> for Database {
+    fn from(inner: D) -> Self {
+        Database {
+            inner: Arc::new(inner),
+            notifications: Arc::new(Notifications::new()),
+        }
+    }
+}
+
+impl Database {
+    pub async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction {
+        DatabaseTransaction::new(
+            self.inner.begin_transaction().await,
+            decoders,
+            self.notifications.clone(),
+        )
+    }
+
+    pub async fn wait_key<K: DatabaseKey + DatabaseKeyPrefixConst>(
```

Comment:
`K` is only for convenience of transforming keys from/to bytes . You could extract a function that works on the bytes only, that doesn't have `K`, then wrap it in a function that only does `K -> bytes` conversion.